File: 3Sum/3Sum.py


class Solution:
    # @return a list of lists of length 3, [[val1,val2,val3]]
    def threeSum(self, num):
        n=len(num)
        num.sort()
        ans=[]
        for i in range(n):
            if i>0 and num[i]==num[i-1]:
                continue
            start=i+1
            end=n-1
            a=num[i]
            while start<end:
                b=num[start]
                c=num[end]
                if start>i+1 and b==num[start-1]:
                    start+=1
                    continue
                if end<n-1 and c==num[end+1]:
                    end-=1
                    continue
                temp=a+b+c
                if temp==0:
                    ans.append([a,b,c])
                    start+=1
                elif temp>0:
                    end-=1
                else:
                    start+=1
        return ans
 

File: 3Sum/3Sum.cpp

class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        sort(num.begin(),num.end());
        int n=num.size();
        vector<vector<int> > ans;
        for(int i=0;i<n;i++)
        {
            if(i>0 && num[i]==num[i-1])
                continue;
            int start=i+1;
            int end=n-1;
            while(start<end)
            {
                if(start>i+1 && num[start]==num[start-1])
                {
                    start++;
                    continue;
                }
                if(end<n-1 && num[end]==num[end+1])
                {
                    end--;
                    continue;
                }
                int temp=num[i]+num[start]+num[end];
                if(temp<0)
                    start++;
                else if(temp==0)
                {
                    vector<int> vec;
                    vec.push_back(num[i]);
                    vec.push_back(num[start]);
                    vec.push_back(num[end]);
                    ans.push_back(vec);
                    start++;
                }
                else
                    end--;
            }
        }
        return ans;
    }
};


 